drivers/mspi/mspi_esp32.c

```diff
@@ -0,0 +1,690 @@
+		return -EINVAL;
+	}
+
+	/* clean up and prepare SPI hal */
+	for (size_t i = 0; i < ARRAY_SIZE(hal->hw->data_buf); ++i) {
+#ifdef CONFIG_SOC_SERIES_ESP32C6
+		hal->hw->data_buf[i].val = 0;
+#else
+		hal->hw->data_buf[i] = 0;
+#endif
+	}
+
+	const struct mspi_xfer_packet *packet = &xfer->packets[packet_index];
+
+	/* Handle DMA buffer allocation if DMA is enabled and we have data to transfer */
+	if (config->dma_enabled && packet->num_bytes > 0) {
+		dma_len = MIN(packet->num_bytes, config->max_dma_buf_size);
+
+		if (packet->dir == MSPI_TX && packet->data_buf) {
+			if (!esp_ptr_dma_capable((uint32_t *)packet->data_buf)) {
+				LOG_DBG("Tx buffer not DMA capable");
+
+				tx_temp = k_aligned_alloc(4, ROUND_UP(dma_len, 4));
+				if (!tx_temp) {
+					LOG_ERR("Error allocating temp buffer Tx");
+					return -ENOMEM;
+				}
+
+				memcpy(tx_temp, packet->data_buf, dma_len);
+			}
+		} else if (packet->dir == MSPI_RX && packet->data_buf) {
+			if (!esp_ptr_dma_capable((uint32_t *)packet->data_buf) ||
+			    ((int)packet->data_buf % 4 != 0) || (dma_len % 4 != 0)) {
+				LOG_DBG("Rx buffer not DMA capable");
+				/* The rx buffer need to be length of
+				 * multiples of 32 bits to avoid heap
+				 * corruption.
+				 */
+
+				rx_temp = k_aligned_alloc(4, ROUND_UP(dma_len, 4));
+				if (!rx_temp) {
+					LOG_ERR("Error allocating temp buffer Rx");
+					return -ENOMEM;
+				}
+			}
+		}
+	}
+
+	/* keep cs line active until last transmission */
+	trans_config->cs_keep_active = xfer->hold_ce;
+
+	/* Handle command phase if present */
+	if (xfer->cmd_length > 0) {
+		trans_config->cmd = packet->cmd;
+		trans_config->cmd_bits = xfer->cmd_length * 8;
+	}
+
+	/* Handle address phase if present */
+	if (xfer->addr_length > 0) {
+		trans_config->addr = packet->address;
+		trans_config->addr_bits = xfer->addr_length * 8;
+	}
+
+	/* Handle data phase if present */
+	if (packet->num_bytes > 0) {
+		size_t bit_len = packet->num_bytes * 8;
+
+		if (config->dma_enabled) {
+			/* bit_len needs to be at least one byte long when using DMA */
+			bit_len = !bit_len ? 8 : bit_len;
+		}
+
+		if (packet->dir == MSPI_TX) {
+			trans_config->send_buffer = tx_temp ? tx_temp : (uint8_t *)packet->data_buf;
+			trans_config->tx_bitlen = bit_len;
+		} else {
+			trans_config->rcv_buffer = rx_temp ? rx_temp : packet->data_buf;
+			trans_config->rx_bitlen = bit_len;
+		}
+	} else if (config->dma_enabled) {
+		trans_config->tx_bitlen = 8;
+		trans_config->rx_bitlen = 8;
+	}
+
+	/* Configure SPI */
+	spi_hal_setup_trans(hal, hal_dev, trans_config);
+	spi_hal_prepare_data(hal, hal_dev, trans_config);
+
+	/* Send data */
+	spi_hal_user_start(hal);
+
+	k_timeout_t timeout = K_MSEC(xfer->timeout ?: MSPI_TIMEOUT_MS);
+	int64_t start_time = k_uptime_get();
+
+	int res = 0;
+
+	while (!spi_hal_usr_is_done(hal)) {
+		if (k_uptime_get() - start_time > timeout.ticks) {
+			LOG_ERR("Transfer timeout");
+			res = -ETIMEDOUT;
+			goto cleanup;
+		}
+
+		k_yield();
```

Comment:
k_yield() only switches if there is a thread with the same priority ready to run, otherwise it is noop, translating this to a busy loop, would not better to consider a k_sleep() instead?

Reply:
As far as I understand it's better not to wait here for any noticeable period of time. I removed the yield, similar to the spi driver.